app/assets/src/components/views/discovery/mapping/MapPlayground.jsx

```diff
@@ -1,78 +1,153 @@
+      return [null, null];
+    } else {
+      return [lat, lon];
+    }
+  };
+
   updateViewport = viewport => {
     this.setState({ viewport });
   };
 
-  renderMarker = (markerData, index) => {
+  renderMarker = (marker, index) => {
     const { viewport } = this.state;
-    const [lat, lon] = markerData[0].split(",");
-    const pointCount = markerData[1].length;
-    const minSize = 14;
+    const [locationName, markerData] = marker;
+    const pointCount = markerData.items.length;
+    const minSize = 12;
     // Scale based on the zoom and point count (zoomed-in = higher zoom)
     const markerSize = Math.max(
-      pointCount * (get("zoom", viewport) || 3),
+      Math.log(pointCount) / Math.log(1.5) * (get("zoom", viewport) || 3),
       minSize
     );
+    const hoverContent = (
+      <div>
+        <div className={cs.title}>{locationName}</div>
+        <div className={cs.description}>{`Samples: ${pointCount}`}</div>
+      </div>
+    );
+    const popupInfo = {
+      name: locationName,
+      lat: markerData.lat,
+      lon: markerData.lon,
+      markerIndex: index,
+      items: markerData.items
+    };
     return (
       <Marker
         key={`marker-${index}`}
-        latitude={parseFloat(lat)}
-        longitude={parseFloat(lon)}
+        latitude={markerData.lat}
+        longitude={markerData.lon}
       >
         <CircleMarker
           size={markerSize}
-          hoverContent={`${pointCount} sample${pointCount > 1 ? "s" : ""}`}
+          hoverContent={hoverContent}
+          onClick={() => this.openPopup(popupInfo)}
         />
       </Marker>
     );
   };
 
+  openPopup(popupInfo) {
+    const popups = Object.assign([], this.state.popups);
+    popups.push(popupInfo);
+    this.setState({ popups });
+  }
+
+  closePopup(popupInfo) {
+    let popups = Object.assign([], this.state.popups);
+    popups = remove({ markerIndex: popupInfo.markerIndex }, popups);
+    this.setState({ popups });
+  }
+
+  renderPopupBox = popupInfo => {
+    return (
+      <MapPopup
+        anchor="bottom"
+        tipSize={0}
+        latitude={popupInfo.lat}
+        longitude={popupInfo.lon}
+        onClose={() => this.closePopup(popupInfo)}
+        className={cs.dataBox}
+      >
+        <div>
+          <div className={cs.title}>{popupInfo.name}</div>
+          {popupInfo.items.map((sample, i) => (
+            <div className={cs.description} key={`popup-${i}`}>
+              {sample.name}
+            </div>
+          ))}
+        </div>
+      </MapPopup>
+    );
+  };
+
   render() {
     const { mapTilerKey } = this.props;
-    const { locationsToSamples } = this.state;
+    const { locationsToItems, popups } = this.state;
 
     return (
       <BaseMap
-        markers={Object.entries(locationsToSamples)}
         mapTilerKey={mapTilerKey}
-        renderMarker={this.renderMarker}
         updateViewport={this.updateViewport}
+        markers={Object.entries(locationsToItems)}
```

Comment:
Do you need to pass both `popups` and `renderPopup` into `<BaseMap>`? Can you render the popups first and just pass those in as a single prop? Same with renderMarker.

Reply:
👍

Reply:
👍 changed